core/dbt/contracts/graph/unparsed.py

```diff
@@ -652,3 +659,18 @@ def validate(cls, data):
         super(UnparsedGroup, cls).validate(data)
         if data["owner"].get("name") is None and data["owner"].get("email") is None:
             raise ValidationError("Group owner must have at least one of 'name' or 'email'.")
+
+
+def normalize_date(d: Optional[datetime.date]) -> Optional[datetime.datetime]:
+    """Convert date to datetime (at midnight), and add local time zone if naive"""
+    if d is None:
+        return None
+
+    # convert date to datetime
+    dt = d if type(d) == datetime.datetime else datetime.datetime(d.year, d.month, d.day)
+
+    if not dt.tzinfo:
```

Comment:
The Python docs [here](https://docs.python.org/3/library/datetime.html#determining-if-an-object-is-aware-or-naive) say the following:

<img width="400" alt="image" src="https://github.com/dbt-labs/dbt-core/assets/44704949/0e136332-7233-4d7d-b251-277af3e56d64">

Translating those into code gives something like [this](https://github.com/dbt-labs/dbt-core/blob/265e09dc93914a14caf0cb0ee778264e20cf75c8/tests/adapter/dbt/tests/adapter/utils/test_current_timestamp.py#L16-L17).

The slight change suggested below would cover the case where `tzinfo` is not `None` but `dt.tzinfo.utcoffset(dt)` returns `None`.

To fully encode the Python datetime definition, I think we'd need to do something like this:
```suggestion
    if not dt.tzinfo or not dt.tzinfo.utcoffset(dt):
```

Granted, it might "never" happen for a `tzinfo` object from `pytz` to return `None` for the offset, but I was able to hand construct such a `tzinfo` object locally.

Reply:
I'd be happy to tighten up this check if you're willing to open the issue. Sorry we didn't get it in v1.